frcxx_nft/src/lib.rs

```diff
@@ -152,17 +179,49 @@ where
     /// Burn a set of NFTs as the owner
     ///
     /// A burnt TokenID can never be minted again
-    pub fn burn(&mut self, owner: ActorID, token_ids: &[TokenID]) -> Result<u64> {
-        let balance = self.state.burn_tokens(&self.bs, owner, token_ids)?;
+    pub fn burn(&mut self, owner: &Address, token_ids: &[TokenID]) -> Result<u64> {
+        let owner = self.msg.resolve_id(owner)?;
+
+        let balance = self.transaction(|state, bs| {
+            Ok(state.burn_tokens(bs, owner, token_ids, |token_data, token_id| {
+                NFTState::assert_owns_token(token_data, token_id, owner)
+            })?)
+        })?;
+
         Ok(balance)
     }
 
     /// Burn a set of NFTs as an operator
     ///
     /// A burnt TokenID can never be minted again
-    pub fn burn_for(&mut self, operator: ActorID, token_ids: &[TokenID]) -> Result<()> {
-        self.state.operator_burn_tokens(&self.bs, operator, token_ids)?;
-        Ok(())
+    pub fn burn_from(
+        &mut self,
+        owner: &Address,
+        operator: &Address,
+        token_ids: &[TokenID],
+    ) -> Result<u64> {
+        let operator = self.msg.resolve_id(operator)?;
+        let owner = self.msg.resolve_or_init(owner)?;
+
+        let balance = self.transaction(|state, bs| {
+            let owner_map = state.get_owner_data_hamt(bs)?;
+            let account_operator = NFTState::is_account_operator(&owner_map, owner, operator)?;
```

Comment:
checking this outside requires an extra load (same for transfer_from)

alternatively this logic can be written in the predicate which means it will be checked for each token. this is not so bad if the predicate has the internally pre-loaded hamt available.

Reply:
https://github.com/helix-onchain/filecoin/issues/171 will track this here unless it's a blocker for merging this PR

Reply:
Not a blocker. But instead of getting the owner map out and then calling a function, make a new state method `is_account_operator(owner, operator)`